### PR TITLE
Waterfall multiple callback defense

### LINKED ISSUE #815: Defend against multiple callbacks in async.waterfall
This would invalidate the test here:  https://github.com/caolan/async/blob/master/test/test-async.js#L883

It will enable use to simplify the implementation significantly, and improve performance.

### PR DESCRIPTION
**Pull request by aearly**
Closes #815 .

This introduces a breaking change -- previously multiple callbacks would call the rest of the waterfall multiple times, now calling a callback multiple times is an error.

This also fixes a detail related to #696 -- no unnecessary deferrals in waterfall.

### CODE REVIEW

File: lib/waterfall.js [hunk 1]
```diff
@@ -5,28 +5,32 @@ import noop from 'lodash/noop';
...
     cb = once(cb || noop);
     if (!isArray(tasks)) return cb(new Error('First argument to waterfall must be an array of functions'));
     if (!tasks.length) return cb();
+    var taskIndex = 0;
 
-    function wrapIterator(iterator) {
-        return rest(function(err, args) {
+    function nextTask(args) {
+        if (taskIndex === tasks.length) {
+            return cb.apply(null, [null].concat(args));
```
Comment (megawac) [inline lib/waterfall.js]: It's being transpiled so you can do `cb(null, ...args)`

File: lib/waterfall.js [hunk 2]
```diff
@@ -5,28 +5,32 @@ import noop from 'lodash/noop';
...
+        }
+        var task = tasks[taskIndex];
```
Comment (megawac) [inline lib/waterfall.js]: do the postfix here `tasks[taskIndex++]`

**Comment by aearly** [comment]
:(  I just noticed that `dist/async.js` isn't run through babel, just rollup.  I'll have to tweak the build if we want to support more than just ES6 modules.

**Comment by megawac** [comment]
Ah, lets not do that for now in that case

**Pull request merged**